holmes/core/tool_calling_llm.py

```diff
@@ -357,6 +365,97 @@ def truncate_messages_to_fit_context(
+            max_context_size = self.llm.get_context_window_size()
+            maximum_output_token = self.llm.get_maximum_output_token()
+            perf_timing.measure("count tokens")
+
+            if (total_tokens + maximum_output_token) > max_context_size:
+                logging.warning("Token limit exceeded. Truncating tool responses.")
+                messages = self.truncate_messages_to_fit_context(
+                    messages, max_context_size, maximum_output_token
+                )
+                perf_timing.measure("truncate_messages_to_fit_context")
+
+            logging.debug(f"sending messages={messages}\n\ntools={tools}")
+            try:
+                full_response = self.llm.completion(
+                    messages=parse_messages_tags(messages),
+                    tools=tools,
+                    tool_choice=tool_choice,
+                    temperature=0.00000001,
+                    response_format=response_format,
+                    stream=False,
+                    drop_params=True,
+                )
+                perf_timing.measure("llm.completion")
+
+            # catch a known error that occurs with Azure and replace the error message with something more obvious to the user
+            except BadRequestError as e:
+                if (
+                    "Unrecognized request arguments supplied: tool_choice, tools"
+                    in str(e)
+                ):
+                    yield json.dumps({"type": "error", "details": {"msg": "The Azure model you chose is not supported. Model version 1106 and higher required."}})
+                    return
+                raise
+            except Exception:
+                raise
+
+            response_message = full_response.choices[0].message
+            tools_to_call = getattr(response_message, "tool_calls", None)
+            if not tools_to_call:
+
+                (text_response, _) = process_response_into_sections(response_message.content)
+                yield json.dumps({"type": "ai_answer", "details": {"answer": text_response}})
+                return
+
+            messages.append(
+                response_message.model_dump(
+                    exclude_defaults=True, exclude_unset=True, exclude_none=True
+                )
+            )
+
+            perf_timing.measure("pre-tool-calls")
+            with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
+                futures = []
+                for t in tools_to_call:
+                    futures.append(executor.submit(self._invoke_tool, t))
+                    yield json.dumps({"type": "start_tool_calling", "details": {"tool_name": t.function.name, "id": t.id}})
+
+                for future in concurrent.futures.as_completed(futures):
+                    tool_call_result: ToolCallResult = future.result()
+                    tool_calls.append(tool_call_result)
+                    messages.append(tool_call_result.as_dict())
```

Comment:
Please save as_dict output to var and reuse instead of calling it multiple times.